投資状況レポートの機能を作成

katanoZ opened the pull request:
ref #16 
投資状況レポートの機能を作成しました。確認をお願いします。
（自分が作成した全てのプロジェクトに対して、誰がいつどれだけ投資したかを一覧で表示する）
- 投資状況レポートの画面機能、`Admin::InvestmentsController` `#report` `#report_result`を作成
（自分の作成したプロジェクトに投資がされている場合、画面ヘッダにレポート画面へのリンクが表示される）
- レポート画面のフォーム`app/forms/investments_report_form.rb`を作成
- レポート処理のためのサービス`app/services/investments_reporter.rb`を作成
- レポート処理用に、カスタムバリデータ`app/validates/date_range_validator.rb`、investmentモデルのスコープ`for_projects_created_by(user)` `created_within(range)`など作成

Commit:
[add]ユーザテーブルに名前カラムを追加

Commit:
[add]投資状況レポート用Adimn::InvestmentsController#reportの雛形を作成

Commit:
[add]投資状況レポート機能を作成

Admin::InvestmentsController#report→#report_resultで出力
自分が作ったプロジェクトに対して、いつ誰がどれだけ投資したかを確認する

Commit:
[clean]investmrnts_reporterのメソッドを整理

Commit:
[clean]investmrnts_reporterのメソッド修正

Commit:
[clean]投資状況レポートフォームのバリデーション修正

Commit:
[clean]投資状況レポートのコード整理と修正

サービスの中に投資状況レポート機能を隠蔽して、コントローラからサービスクラスのみ見えるようにしたい為。
フォームはサービスから呼び出すよう修正。

katanoZ commented:
一旦、レビュー指摘事項の修正を行いました。
確認をお願いいたします。
- コントローラ（Admin::InvestmentsController）から、サービス（services/InvestmentsReporter）のみ扱うようにコードを修正。（フォーム（InvestmentsReportForm）はサービスから使うよう修正）
- app/validators/date_range_validator.rbを削除して、フォームクラス内に全てのバリデーションを書く形に修正

kabane commented on app/forms/investments_report_form.rb:
```diff
@@ -0,0 +1,34 @@
+    errors.add(attribute, 'は1年前から本日までの期間で入力してください')
+  end
+
+  def start_and_end_dates_entered?
```
@katanoZ この処理はverify_rangeの中で行ってしまって良い気がしますー。

katanoZ replied:
確かに、その方が見やすい（書き方に統一性がある？）と思いました。
修正対応します。

kabane commented on app/forms/investments_report_form.rb:
```diff
@@ -0,0 +1,34 @@
+  end
+
+  def verify_date_term(date, attribute)
+    return if date.between?(Date.today.prev_year, Date.today)
```
@katanoZ ここですが、やはり「未来の日付」での期間指定は許可しても良い気がしてきました。未来の日付を指定した場合、該当するデータが表示されないだけなので、わざわざバリデーションで弾くほどのことないのかなぁと。なので、ここは消しでお願いできればとー。

katanoZ replied:
確かに「未来の日付」は、入力した日付の論理的な妥当性としては特に問題ないので、バリデーションの役割としては不適当だと思いました。（もしやりたいのなら、画面で入力を制限するくらいが適当かと思いました。）
修正対応します。

Commit:
[add]投資状況レポートのCSV出力機能を追加

Commit:
[update]投資状況レポートフォームのバリデーションを修正

未来日を入力させないバリデーションは、サービスのバリデーションとして不適当なため削除
未来の日付は論理的には妥当な入力値であり、未来の日付を入力しても該当のデータがないだけなので、バリデーションとしては扱わない

katanoZ commented:
以下の対応を行いましたので確認をお願いいたします。
- レビュー指摘事項の修正対応（投資状況レポートフォームのバリデーション修正）
- 投資状況レポートのCSV出力機能を追加

CSV出力機能について
- 投資状況レポートが画面に正常に表示された場合、画面に「CSV出力」のリンクが表示される
（画面 `app/views/admin/investments/report.html.slim`）
- リンク先のコントローラ`InvestmentsController#generate_csv_report`アクションでCSVを出力
- CSV生成処理はサービス`InvestmentsReporter`の`generate_csv`メソッドで行う

kabane commented on app/services/investments_reporter.rb:
```diff
@@ -0,0 +1,53 @@
+          investment.project.name,
+          investment.user.name,
+          investment.price.to_s(:currency),
+          I18n.l(investment.created_at, format: :long)
```
@katanoZ created_atは不要ですー。

katanoZ replied:
created_atは削除します。

kabane commented on app/services/investments_reporter.rb:
```diff
@@ -0,0 +1,53 @@
+              .order(:project_id, :user_id)
+  end
+
+  def generate_csv
```
@katanoZ 
メソッド名は、export_csvの方が妥当かと。
あとこのロジックをモデルに持たせるかは、悩ましいところですねー。
libに新たにutilsというフォルダを作成し、CSVのエクスポート機能はそちらに移したいですねー。

実行のトリガーとなるメソッド自体は、InvestmentsReporter#csv_exportで問題ないのですが、CSVを作成するロジックをutilsに移したい感じです！

katanoZ replied:
lib/utilsにCSV生成部分のコードを移動するように修正します。

Commit:
[clean]投資状況レポートサービスのCSV出力メソッド名を修正

Commit:
[update]投資状況レポートのCSV生成機能をlib/utilsに移動

katanoZ commented:
投資状況レポートのCSV生成機能をlib/utilsに移動しました。
確認をお願いいたします。

kabane commented on app/services/investments_reporter.rb:
```diff
@@ -0,0 +1,44 @@
+    @user = user
+  end
+
+  def search
```
@katanoZ すいません、見落としておりましたが、ここはサーチではなくinvestmentsで良い気がしましたー。

katanoZ replied:
修正します！

Commit:
[clean]投資状況レポートサービスのメソッド名を整理
